### ISSUE TITLE
Slow Seek/Read Performance

### ISSUE THREAD

**Issue by animosity22** [ncw/rclone#1298]
I'm trying to figure out why the seek is so slow when it tries to read a file.
In this example, it's a really small 225MB file that it tries to do a seek on to get some info for a plex media scan.

I have a 750/750 FIOS line and when this running, I barely see any utilization on my line at maybe 5-10Mbs. If I play the same file, it seems to play fine with a few seconds to download and my line pops up to 200-250Mbs to grab the file and put it in the buffer.

> What is your rclone version (eg output from `rclone -V`)
```
rclone v1.36-06-g4dc030dβ
```
> Which OS you are using and how many bits (eg Windows 7, 64 bit)
```
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 8.7 (jessie)
Release:	8.7
Codename:	jessie
Linux plex 3.16.0-4-amd64 #1 SMP Debian 3.16.39-1+deb8u2 (2017-03-07) x86_64 GNU/Linux
```
> Which cloud storage system are you using? (eg Google Drive)
ACD

> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
```
rclone mount \
--read-only \
--allow-other \
--default-permissions \
--uid 1000 \
--gid 1000 \
--umask 002 \
--acd-templink-threshold 0 \
--buffer-size 1G \
--max-read-ahead 32M \
--syslog \
--stats 1m \
-vv \
media: /media &
```

This is a crypt mount as well.

> A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
My scan looks like:

```
"TV Scan"
Mar 29 19:42:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: Dir.Lookup
Mar 29 19:42:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: Dir.Lookup OK
Mar 29 19:42:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Attr valid=1m0s ino=0 size=235111697 mode=-rw-rw-r--
Mar 29 19:42:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Open OpenReadOnly
Mar 29 19:42:34 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 4096 offset 235110400
Mar 29 19:42:34 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 0 to 235110400 (io.Seeker)
Mar 29 19:42:35 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 29 19:42:35 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 16384 offset 0
Mar 29 19:42:35 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 235111697 to 0 (io.Seeker)
Mar 29 19:42:35 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 29 19:42:35 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 4096 offset 234377216
Mar 29 19:42:35 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 16384 to 234377216 (io.Seeker)
Mar 29 19:42:36 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 29 19:42:36 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 4096 offset 234831872
Mar 29 19:42:36 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 234381312 to 234831872 (io.Seeker)
Mar 29 19:42:36 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 29 19:43:02 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush
Mar 29 19:43:02 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush OK
Mar 29 19:43:02 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release closing
Mar 29 19:43:02 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release OK
```
It takes about 35-45 second per file that it runs through.

and when I go to play I see:
```
Mar 30 06:16:08 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: Dir.Lookup
Mar 30 06:16:08 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: Dir.Lookup OK
Mar 30 06:16:08 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Attr valid=1m0s ino=0 size=235111697 mode=-rw-rw-r--
Mar 30 06:16:08 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Open OpenReadOnly
Mar 30 06:16:14 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush
Mar 30 06:16:14 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush OK
Mar 30 06:16:14 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release closing
Mar 30 06:16:14 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release OK
Mar 30 06:16:14 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Open OpenReadOnly
Mar 30 06:16:17 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 0
Mar 30 06:16:18 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:18 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 131072
Mar 30 06:16:18 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:18 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Open OpenReadOnly
Mar 30 06:16:22 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 65536 offset 234356736
Mar 30 06:16:22 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 0 to 234356736 (io.Seeker)
Mar 30 06:16:24 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush
Mar 30 06:16:24 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush OK
Mar 30 06:16:24 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release closing
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 234422272
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 234553344
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 234684416
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 234815488
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 234946560
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 36864 offset 235077632
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release closing
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release OK
Mar 30 06:16:25 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Open OpenReadOnly
Mar 30 06:16:29 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 65536 offset 70385664
Mar 30 06:16:29 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 0 to 70385664 (io.Seeker)
Mar 30 06:16:30 plex rclone[12415]: #012Transferred:   5.942 GBytes (148.554 kBytes/s)#012Errors:                 0#012Checks:                 0#012Transferred:         6946#012Elapsed time:  11h39m0.7s#012Transferring:#012 * ...ler Dealers/Wheeler.Dealers.S03E11.DVD.mp4:  0% done, 0 Bytes/s, ETA: -
Mar 30 06:16:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 70451200
Mar 30 06:16:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 70582272
Mar 30 06:16:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:33 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Open OpenReadOnly
Mar 30 06:16:37 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 65536 offset 145391616
Mar 30 06:16:37 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 0 to 145391616 (io.Seeker)
Mar 30 06:16:40 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush
Mar 30 06:16:40 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush OK
Mar 30 06:16:40 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release closing
Mar 30 06:16:42 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:42 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 145457152
Mar 30 06:16:42 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:42 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Open OpenReadOnly
Mar 30 06:16:44 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 65536 offset 178618368
Mar 30 06:16:44 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 0 to 178618368 (io.Seeker)
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 178683904
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush OK
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release closing
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: File.Open OpenReadOnly
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Flush OK
Mar 30 06:16:48 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Release closing
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 65536 offset 65536
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.seek from 0 to 65536 (io.Seeker)
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 131072
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 262144
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 393216
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 524288
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 655360
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 786432
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 917504
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read OK
Mar 30 06:16:53 plex rclone[12415]: TV/Wheeler Dealers/Wheeler.Dealers.S03E11.DVD.mp4: ReadFileHandle.Read size 131072 offset 1048576
```

And it continues to read the file during the course of the play and streaming works without a problem.

**Comment by ncw** [comment]
Seeking is slow because rclone has to close the existing stream, and re-open it using a `Range:` header.

There is a test script in `cmd/mount/seek_speed.go` which you can use to test the seek speeds.

Here is me running it on a mounted encrypted acd disk

You can see seeks are taking on average 2 seconds for me.

```
$ go run seek_speed.go ~/mnt/tmp/bigfile
2017/03/31 13:16:42 Reading 170511 from 150060860
2017/03/31 13:16:43 Reading 459963 from 139102236
2017/03/31 13:16:47 Reading 643462 from 1819658102
2017/03/31 13:16:48 Reading 578220 from 1654194378
2017/03/31 13:16:49 Reading 280484 from 1769279246
2017/03/31 13:16:49 Reading 350895 from 810238284
2017/03/31 13:16:55 Reading 614385 from 967797739
2017/03/31 13:16:55 Reading 300314 from 1621112450
2017/03/31 13:16:57 Reading 285845 from 1112190248
2017/03/31 13:16:57 Reading 625634 from 1595138255
2017/03/31 13:16:58 Reading 67802 from 336933638
2017/03/31 13:16:58 Reading 401042 from 1216435296
2017/03/31 13:16:59 Reading 445739 from 2034103843
2017/03/31 13:17:04 Reading 85240 from 1090106652
2017/03/31 13:17:10 Reading 923639 from 2185605778
2017/03/31 13:17:10 Reading 1011576 from 1289722284
2017/03/31 13:17:11 Reading 686863 from 1219131873
2017/03/31 13:17:15 Reading 756044 from 235146038
2017/03/31 13:17:22 Reading 82935 from 73103248
2017/03/31 13:17:23 Reading 543634 from 1037027997
2017/03/31 13:17:27 Reading 180882 from 502230445
2017/03/31 13:17:31 Reading 402243 from 1873257165
2017/03/31 13:17:31 Reading 670611 from 1972506627
2017/03/31 13:17:33 Reading 735972 from 463388514
2017/03/31 13:17:34 Reading 1025113 from 1917938230
2017/03/31 13:17:35 That took 52.400670551s for 25 iterations, 2.096026822s per iteration
```

**Comment by animosity22** [comment]
Yeah, after posting, I came to the conclusion that seeks seems to be working as expected, but I was hoping it would use up more of my bandwidth to be faster.

The root of my problem seems to be some problematic items that won't get the metadata properly updated so Plex keeps kicking off seeks of that file over and over again after the last scan. I have another library setup all in ACD via a crypt / clone mount and that works fine. 

With the debugging, I can see the same files being inspected over and over again.

If it helps folks, I did post on the plex forums to try to identify the cause but have yet to find an answer.

http://forums.plex.tv/discussion/comment/1408733#Comment_1408733
